brushfire-core/src/main/scala/com/stripe/brushfire/Tree.scala

```diff
@@ -31,6 +32,107 @@ case class Tree[K, V, T](root: Node[K, V, T]) {
     }
   }
 
+  /**
+   * Prune a tree to minimize validation error.
+   * Replaces a split with a leaf when it would have a lower error than the child leaves.
+   *
+   * @param validationData
+   * @param start
+   * @param voter
+   * @param error
+   * @param weightMonoid
+   * @param errorOrdering
+   * @tparam P
+   * @tparam E
+   * @return
+   */
+  def prune[P, E](validationData: Map[Int, T], start: Option[Node[K, V, T]] = None, voter: Voter[T, P], error: Error[T, P, E])(implicit weightMonoid: Monoid[T], errorOrdering: Ordering[E]): Tree[K, V, T] = {
+    val begin: Node[K, V, T] = start.getOrElse(this.root) // Use root node as default root.
+    Tree(pruneNode(validationData, begin, voter, error))
+      .growByLeafIndex { i => Nil } //this will renumber all the leaves
```

Comment:
IIRC this is the second place this pattern is used, in which case maybe we should have a `renumberLeaves` method to make it more clear?